csrc/adam/cpu_adam_impl.cpp

```diff
@@ -112,37 +148,65 @@ void Adam_Optimizer::Step_1(float* _params,
                 _buf_index = !_buf_index;
             }
 #endif
+            }
+        ret = std::max(ret, std::make_pair(
+                                *std::max_element(tmp_max_avg, tmp_max_avg + OMP_MAX_THREADS),
+                                *std::max_element(tmp_max_var, tmp_max_var + OMP_MAX_THREADS)));
         }
     }
+    return ret;
 }
 
-void Adam_Optimizer::Step_4(float* _params,
+std::pair<float, float> Adam_Optimizer::Step_4(float* _params,
                             float* grads,
                             float* _exp_avg,
                             float* _exp_avg_sq,
                             size_t _param_size,
                             ds_half_precision_t* dev_params,
-                            bool half_precision)
+                            bool half_precision_param,
+                            bool half_precision_opt_state,
+                            float avg_scaling,
+                            float avg_sq_scaling, 
+                            float new_avg_scaling, 
+                            float new_avg_sq_scaling)
 {
     size_t rounded_size = 0;
+    std::pair<float, float> ret = std::make_pair(0.0f, 0.0f);
+
 #if defined(__AVX512__) or defined(__AVX256__)
-    Step_AVX<4>(&rounded_size,
+    
+    auto ret_avx = Step_AVX<4>(&rounded_size,
                 _params,
                 grads,
                 _exp_avg,
                 _exp_avg_sq,
                 _param_size,
                 dev_params,
-                half_precision);
+                half_precision_param,
+                half_precision_opt_state,
+                avg_scaling,
+                avg_sq_scaling, 
+                new_avg_scaling, 
+                new_avg_sq_scaling);
+    ret = std::max(ret, ret_avx);
+
 #endif
-    if (_param_size > rounded_size)
-        Step_1((_params + rounded_size),
-               (grads + rounded_size),
-               (_exp_avg + rounded_size),
-               (_exp_avg_sq + rounded_size),
+    if (_param_size > rounded_size) {
+        auto ret_remain = Step_1((_params + (rounded_size >> static_cast<int> (half_precision_param))),
+               (grads + (rounded_size >> static_cast<int> (half_precision_param))),
+               (_exp_avg + (rounded_size >> static_cast<int> (half_precision_opt_state))),
+               (_exp_avg_sq + (rounded_size >> static_cast<int> (half_precision_opt_state))),
```

Comment:
can we please move this part of shifting to a separate utility function?

Reply:
Yes. I can write a define for this.